[2단계 - 장기] 강산(조재중) 미션 제출합니다.

m-a-king opened the pull request:
<!-- 

## 코드 리뷰 팁

- 코드와 관련된 질문이 있다면, PR 본문에 적기 보다는 해당 코드를 선택하고 코멘트를 남겨주세요.
  - [참고: Adding comments to a pull request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request#adding-comments-to-a-pull-request)

-->

## 체크 리스트

- [x] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x] 애플리케이션이 정상적으로 실행되나요?
- [x] [프롤로그](https://prolog.techcourse.co.kr)에 셀프 체크를 작성했나요?
  - <!-- 작성한 셀프 체크의 링크를 남겨주세요. -->
https://prolog.techcourse.co.kr/studylogs/4107
> 
> ## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?
> 
> ### 1~5점 중에서 선택해주세요.
> 
> - [ ] 1 (전혀 충족하지 못함)
> - [ ] 2
> - [ ] 3 (보통)
> - [x] 4
> - [ ] 5 (완벽하게 충족)
> 
> ### 선택한 점수의 이유를 적어주세요.
> 데이터베이스 연동과 함께 추가된 코드들을 깔끔히 정리하지 못했습니다.
> 
> ## 어떤 부분에 집중하여 리뷰해야 할까요?
> 
> <!-- 리뷰어가 효과적으로 피드백할 수 있도록 중점적으로 피드백받고 싶은 내용을 공유해주세요.  
> 예를 들어, 가장 고민했던 점이나 여전히 어려운 부분, 그리고 이에 대한 생각을 적을 수 있습니다. -->
> 
> 1. 데이터베이스 기능이 추가되면서 기존 Janggi 클래스가 입출력 뷰와 DB 통신까지 모두 담당하게 되어, 클래스의 책임이 무거워지고 가독성도 떨어지는 문제가 있었습니다. 이를 해결하기 위해 Service 계층과 Repository 계층을 도입하여 역할을 분리했습니다.
> 
> 2. 별도의 DAO 또는 테이블 구조에 대응되는 전용 객체는 만들지 않았습니다. 통신 효율을 고려하면 DAO(또는 DB 전용 모델 객체)가 필요할 수도 있다고 생각했지만, 현재는 자동 저장 기능도 없고, 기물의 이동 경로도 저장하지 않으며, DB와의 통신 빈도 자체가 많지 않은 상황이라 판단하여 별도로 두지 않았습니다. 통신 횟수가 다소 늘어나는 부분이 있지만, 감당 가능한 수준이라고 생각했습니다.
> 현재 SQL 처리는 Repository의 구현체인 `MysqlRepository` 클래스 내부에서 직접 수행하고 있으며, 이 구조는 현재 프로젝트의 규모나 복잡도를 고려했을 때 과하지 않고 적절하다고 판단했습니다. 다만, 계층 간 책임이 일부 침범되고 있는 부분은 추후 리팩토링을 통해 개선할 예정이며, 기능이 확장되거나 구조가 복잡해질 경우에는 DAO 혹은 DB 전용 객체 도입을 고려할 계획입니다.
> 
> 3. (2.1) 궁성 구현까지는 테스트 코드와 함께 개발을 진행했으나, (2.2) DB 연동부터 시간 제약으로 인해, 별도의 테스트 코드 없이 실제 프로그램을 실행하며 진행했습니다. 추가된 코드들에 대한 테스트를 작성하고, 리팩토링도 함께 진행할 계획입니다.
> 
> 
> ---
> 
> # 1단계에서 이어지는 질문
> 
> ```
> 베루스: 
> 정적 팩터리 메서드는 충분히 얘기를 나눈 것 같네요.
> 추가로 아래의 내용만 더 얘기를 나눠보고 싶네요.
> 
> 필드의 타입과 생성자의 파라미터 타입을 일치시키는 것입니다.
> 
> 필드 타입과 생성자의 파라미터 타입을 일치시키게 되면 외부에서 객체의 내부 자료구조를 다 알게 되는거 아닌가요??
> ```
> 
> https://github.com/woowacourse/java-janggi/pull/67#discussion_r2018055702
> 
> 
> <!-- This is an auto-generated comment: release notes by coderabbit.ai -->
> 
> ## Summary by CodeRabbit
> 
> - **새로운 기능**
> 	- 플레이어가 게임 진행 상황을 저장하고, 종료 후에도 이어서 플레이할 수 있는 게임 저장 기능 도입
> 	- 궁 관련 기물 이동 규칙이 업데이트되어, 기물의 이동 경로와 제한이 향상됨
> 	- 게임 시작 시 신규 게임 생성, 저장 게임 불러오기, 종료 옵션 선택 인터페이스가 개선됨
> 
> - **백엔드 개선**
> 	- 안정적인 게임 상태 관리를 위한 데이터베이스 스키마 및 연동 시스템이 강화됨
> 
> <!-- end of auto-generated comment: release notes by coderabbit.ai -->
> 
> 

## 어떤 부분에 집중하여 리뷰해야 할까요?

1. JanggiService의 서비스 계층 설계
    - 서비스에서 트랜잭션 처리를 위임받아 사용하는 흐름과 책임 분리가 자연스러운지 확인 부탁드립니다.

2. 커맨드 패턴 적용 구조

3. TestContainer 기반 테스트 환경
    - 기존 H2의 한계(upsert vs merge) 때문에 실제 동작과 다른 문제가 있었고, 이를 보완하기 위해 MySQL TestContainer를 도입했습니다.

4. 코멘트에 추가했습니다!

m-a-king committed:
docs: 궁성 규칙 추가

m-a-king committed:
feat: 대각선 확인 기능 추가

m-a-king committed:
feat: Vector 방향 체이닝 메서드 추가

m-a-king committed:
feat: 궁성 추가

m-a-king committed:
refactor: 가독성 개선

m-a-king committed:
feat: 궁성 내 단일 움직임 검증 추가

m-a-king committed:
feat: 궁성 내 직선 움직임 검증 추가

m-a-king committed:
feat: 궁과 사의 궁성 외부 움직임 제약 추가

m-a-king committed:
refactor: 예외 메시지 개선

m-a-king committed:
refactor: 각 팀의 피스들의 행의 합 변수명 개선

m-a-king committed:
refactor: 외부에서 사용하지 않는 메서드 접근 제어자 변경, 가독성 개선

m-a-king committed:
test: 궁성 구현으로 기물 좌표 수정

m-a-king committed:
refactor: 메서드 추출, 가독성 개선

m-a-king committed:
refactor: 기물 삭제 책임 이동, 가독성 개선

m-a-king committed:
refactor: final 키워드 추가

m-a-king committed:
refactor: 장애물 검사를 움직임 검사에서 함께 진행

m-a-king committed:
refactor: 가독성 개선

m-a-king committed:
refactor: 사용하지 않는 메서드 삭제

m-a-king committed:
test: 행, 열 기본 생성 테스트 추가

m-a-king committed:
test: 궁, 사의 궁성 제한 테스트 추가

m-a-king committed:
test: 각종 예외 테스트 추가

m-a-king committed:
refactor: 가독성 개선

m-a-king committed:
refactor: 패키지 이동

m-a-king committed:
feat: 정적 팩토리 메서드 추가

m-a-king committed:
docs: 게임 저장을 위한 데이터베이스 연동 준비

m-a-king committed:
feat: db connection provider 추가

m-a-king committed:
feat: piece 정적 팩토리 메서드 추가, 심플 piece 팩토리 추가

m-a-king committed:
feat: turn getter 추가

m-a-king committed:
feat: inputView 개선

m-a-king committed:
refactor: 게임 종료 클래스 명명 개선

m-a-king committed:
feat: game, piece repository 인터페이스 추가

m-a-king committed:
feat: 게임 저장/불러오기/종료 기능 추가

m-a-king committed:
feat: repository 구현체 추가

m-a-king committed:
feat: main 의존성 주입 수정

verus-j commented on src/main/java/janggi/board/Palace.java:
```diff
@@ -0,0 +1,64 @@
+        return this.team == team;
+    }
+
+    public List<Position> getPositions() {
+        return Collections.unmodifiableList(positions);
+    }
```
사용하지 않는 메서드네요~

m-a-king replied:
네, 삭제하겠습니다!

verus-j commented on src/main/java/janggi/board/Board.java:
```diff
@@ -72,4 +83,8 @@ private void updateBoard(final Position departure, final Position destination, f
         return Collections.unmodifiableMap(positionToPiece);
     }
+
+    public Pieces getAlivePieces() {
```
테스트에서 사용되지 않는 것 같은데, 검증문을 작성할 때 활용할 수 있지 않을까요?

m-a-king replied:
추가했습니다!

verus-j commented on src/main/java/janggi/piece/Pieces.java:
```diff
@@ -32,6 +39,12 @@ public Pieces addAll(final Pieces pieces) {
         return addAll(pieces.getPieces());
     }
 
+    public Pieces getByTeam(final Team team) {
```
테스트가 없네요~

m-a-king replied:
추가했습니다!

verus-j commented on src/main/java/janggi/player/Turn.java:
```diff
@@ -24,4 +24,8 @@ public Team getCurrentTeam() {
         accumulatedCount++;
     }
+
+    public int getAccumulatedCount() {
```
여기도 테스트 코드에서 검증문을 작성할 때 활용할 수 있을 것 같습니다~

m-a-king replied:
넵, 감사합니다!

verus-j commented on src/main/java/janggi/Janggi.java:
```diff
@@ -1,47 +1,114 @@
-        Turn turn = Turn.start();
-        Players players = Players.create(turn);
-        Board board = players.createBoard();
+    private GameContext prepareGameContext() {
```
isQuit나 마지막 RuntimeException을 던지는 코드를 봤을 때 GameContext를 준비하는 것 외의 동작들도 있는 것 같습니다. GameQuitException은 예외 상황을 처리하는 것보다는 제어 흐름을 위해 사용하는 것 같습니다. 이 부분을 고민해서 개선하면 좋을 것 같아요~
참고: https://m.blog.naver.com/fbfbf1/223076437697

m-a-king replied:
의도적인 예외라면 괜찮을 줄 알았는데, 수정하고 보니 말씀하신 대로 확실히 예외 상황에서만 예외를 다루는 것이 좋을 것 같습니다.

verus-j commented on src/main/java/janggi/GameContext.java:
```diff
@@ -0,0 +1,52 @@
+        this.turn = turn;
+    }
+
+    public void setGameId(final long gameId) {
```
여기는 사용하지 않는 메서드네요

m-a-king replied:
삭제했습니다!

verus-j commented on src/main/java/janggi/GameContext.java:
```diff
@@ -0,0 +1,52 @@
+        return gameId;
+    }
+
+    public LocalDateTime getStartAt() {
```
여기도 사용하지 않는 메서드입니다

m-a-king replied:
삭제했습니다!

verus-j commented on src/main/java/janggi/service/JanggiService.java:
```diff
@@ -0,0 +1,82 @@
+        final Pieces pieces = Pieces.from(piecesList);
+        final Turn turn = gameRepository.findTurnByGameId(selectGameId)
+                .orElseThrow(() -> new NoSuchElementException("저장된 게임이 없습니다"));
+        final LocalDateTime startAt = null;
```
여기서 null은 어떤 의미인가요?

m-a-king replied:
죄송합니다... 제출 마감을 맞추느라, 올바른 코드를 작성하지 못한 것 같습니다 ㅠㅠ

수정했습니다!

verus-j commented on src/main/java/janggi/Janggi.java:
```diff
@@ -1,47 +1,114 @@
+                            player,
+                            (MoveCommand) command);
+                    checkWinner(players);
+                }
 
-                board.movePiece(
-                        player,
-                        moveCommand.getDeparturePosition(),
-                        moveCommand.getDestinationPosition());
+                if (command.getType().isSave()) {
+                    janggiService.saveGame(
+                            gameId,
+                            players,
+                            turn,
+                            board.getAlivePieces().getPieces());
+                    return;
+                }
 
-                processScore(players);
-            } catch (GameOverException e) {
+                if (command.getType().isQuit()) {
+                    throw new GameQuitException();
+                }
```
GameOverException을 던지는 부분도 앞에서 언급했듯이 제어 흐름을 위해 사용하는 것 같습니다. 예외는 예외 상황에서만 사용하도록 수정하는건 어떨까요?

m-a-king replied:
맞습니다! 수정했습니다.

verus-j commented on src/main/java/janggi/service/JanggiService.java:
```diff
@@ -0,0 +1,82 @@
+        return false;
+    }
+
+    public void saveGame(Long gameId, final Players players, final Turn turn, final List<Piece> pieces) {
```
저장하지 않은 게임을 여러번 save 할 경우 gameId는 항상 null이여서 각각 다른 gameId를 가진 게임을 저장하지 않나요?

m-a-king replied:
맞습니다. 개선된 현재 코드도 동일 게임을 여러 번 저장하게 되면 말씀하신 문제가 발생할 수 있습니다.
다행히 현재는 저장 후 바로 게임을 종료시키기 때문에 반복 저장이 발생할 수 없지만, 구조적으로는 여전히 위험 요소가 존재합니다.
이 문제는 save 시 반환된 GameId로 GameContext를 업데이트하는 방식으로 해결할 수 있을 것 같습니다.

verus-j commented on src/main/java/janggi/player/Score.java:
```diff
@@ -34,11 +34,15 @@ public static Score win() {
     }
 
-    public Score add(Score score) {
+    public Score add(final Score score) {
```
이전 단계의 코드여서 코멘트를 달 수 없어서 여기 남깁니다. general의 경우 19980608이라는 값을 사용하는데 정상적인 점수일까요...?

m-a-king replied:
저도 값을 하드코딩 해뒀다가 잊었네요.. 수정했습니다!

verus-j commented on src/main/java/janggi/service/JanggiService.java:
```diff
@@ -0,0 +1,82 @@
+        board.movePiece(player, command.getDeparturePosition(), command.getDestinationPosition());
+    }
+
+    public boolean isGameOver(final Players players) {
+        for (final Team team : Team.values()) {
+            if (players.getScore(team).isGreaterThan(Score.win())) {
+                return true;
+            }
+        }
+        return false;
+    }
```
여기는 굳이 JanggiService가 아니여도 Players에서 처리할 수 있을 것 같습니다

m-a-king replied:
감사합니다. 수정했습니다!

verus-j commented on src/main/java/janggi/Janggi.java:
```diff
@@ -1,47 +1,114 @@
+                    janggiService.movePiece(
+                            board,
+                            player,
+                            (MoveCommand) command);
```
MoveCommand로 명시적으로 타입을 변경하는 이유가 있을까요?

m-a-king replied:
당시에는 Command가 명확하게 정립되지 않은 상태였고, MoveCommand의 기능을 사용하기 위해 다운캐스팅이 필요했습니다.
현재는 커맨드 패턴으로 구조를 리팩토링하면서 해결했습니다!

m-a-king committed:
test: 보드, 기물 테스트 추가

verus-j commented on src/main/java/janggi/piece/General.java:
```diff
@@ -28,6 +30,14 @@ protected Piece createPiece(final Position position) {
 
+    @Override
+    protected void validateSpecialRule(final Board board, final Position destination) {
+        if (board.isPalace(destination)) {
```
이전 단계에서는 SingleMovementRule은 한칸 이동에만 관심을 두고 세세한 개별 규칙은 SpecialRule에서 다뤘는데, 지금은 궁성에 대한 특별한 규칙도 SingleMovementRule에서 다루고 있네요. 공통된 관심사가 생긴 것 같은 느낌입니다. 한 쪽으로 합쳐서 관심사를 다시 분리하는게 좋을 것 같아요~

m-a-king replied:
여러 번 롤백하며 전략을 동적으로 변경하기, 상속 구조 등 다양한 방향으로 고민해봤지만, 코드 구조를 크게 변경하기 어려운 상황에서 MovementRule 구현체 내부에 궁성 관련 규칙을 필드로 두는 방식이 가장 현실적인 선택이라 판단했습니다. 피드백 부탁드리겠습니다!

verus-j replied:
궁성관련 로직을 한 곳을 잘 모은것 같습니다! 이전보다 한눈에 코드를 파악할 수 있게 된 것 같아요~

verus-j commented on src/main/java/janggi/piece/rule/movement/StraightMovementRule.java:
```diff
@@ -0,0 +1,59 @@
+    protected void validateMoveShape(final Board board, final Position departure, final Position destination) {
+        final Distance distance = Distance.of(departure, destination);
+
+        if (distance.isStraight()) {
+            return;
+        }
+
+        if (isValidPalaceDiagonalMove(board, departure, destination, distance)) {
+            return;
+        }
```
여기 조건을 하나로 합쳐도 좋을 것 같습니다~

m-a-king replied:
가독성을 위해서 일부러 나눴었는데, 피드백을 듣고 다시 보니 합쳐진 것이 더 좋은 것 같습니다!

verus-j commented on src/main/java/janggi/piece/rule/movement/StraightMovementRule.java:
```diff
@@ -0,0 +1,59 @@
+                                              final Position departure,
+                                              final Position destination,
+                                              final Distance distance) {
+        if (!board.isPalace(departure)) {
+            return false;
+        }
+
+        if (!board.isPalace(destination)) {
+            return false;
+        }
+
+        if (!distance.isDiagonal()) {
+            return false;
+        }
```
여기도 조건을 하나로 합쳐도 될 것 같아요~ 이외에도 다른 Rule에서도 합칠 수 있는 조건문을 확인해보면 좋을 것 같아요~

m-a-king replied:
네, 수정했습니다!

verus-j reviewed (changes requested):
안녕하세요 강산~
부족한 시간임에도 2단계 진행하시느라 고생 많으셨습니다!
피드백 남겨드렸으니 확인부탁드려요~

verus-j commented on DDL.sql:
```diff
@@ -0,0 +1,23 @@
+CREATE TABLE game
```
https://github.com/woowacourse/java-janggi/pull/67#discussion_r2021009002

답글도 남겨뒀습니다~

m-a-king replied:
https://github.com/woowacourse/java-janggi/pull/67#discussion_r2027406361

확인했습니다! 감사합니다.

m-a-king committed:
refactor: import 정리

m-a-king committed:
feat: 정적 팩토리 메서드 추가

m-a-king committed:
docs: DDL 수정

m-a-king committed:
refactor: 사용하지 않는 메서드 삭제

m-a-king committed:
refactor: Janggi 클래스 개선 (커맨드 패턴 추가 등)

m-a-king committed:
feat: ResultSet 0개 이상의 데이터를 리스트로 변환하는 유틸 추가

m-a-king committed:
feat: Transaction 추가

m-a-king committed:
feat: 테이블과 매핑되도록 DB 조회 수정

- 테이블과 1:1 매핑되는 DTO 클래스 생성
- Connection을 repository의 필드에서 제거하고, 메서드 인자로 전달되도록 변경

m-a-king committed:
feat: 트랜잭션 관리를 위해서 서비스에서 DB 커넥션 관리, 전체적인 로직 개선

m-a-king committed:
feat: 저장된 게임이 없는 경우 처리 추가

m-a-king committed:
feat: 기물 삭제 후, 저장하도록 수정

m-a-king committed:
refactor: Players 관련 기능을 GameContext에서 제공하도록 위임

m-a-king committed:
refactor: 불필요한 로직 삭제, 메서드명 개선

m-a-king committed:
fix: 출력 문제 해결

m-a-king committed:
refactor: command 출력 관심사 제거

m-a-king committed:
test: Command 구현체 테스트 추가

m-a-king committed:
test: Players, Player 승리 확인 테스트 추가

m-a-king committed:
test: GameContext 테스트 추가

m-a-king committed:
test: GameId 테스트 추가

m-a-king committed:
test: Transaction 테스트 추가

m-a-king committed:
test: Repository 테스트 추가, TestContainer 의존성 설정

- H2에서 upsert 미지원 문제 해결을 위해 TestContainer 사용

m-a-king committed:
test: Service 테스트 추가

m-a-king committed:
feat: 게임 상태 변경 기능 추가

m-a-king committed:
refactor: 궁성 외부로 향하는 움직임 제한 로직 추출

m-a-king committed:
refactor: 궁성 관심사 분리

verus-j reviewed (approved):
안녕하세요 강산~
레벨1 진행하시느라 고생 많으셨습니다
피드백 잘 반영해주셔서 여기서 이만 머지합니다~
방학 푹 쉬시고 레벨2도 화이팅입니다!
